mlem/contrib/kubernetes/base.py

```diff
@@ -0,0 +1,126 @@
+import os
+from typing import ClassVar, Optional
+
+from kubernetes import config
+
+from mlem.config import project_config
+from mlem.core.objects import (
+    DeployState,
+    DeployStatus,
+    MlemDeployment,
+    MlemEnv,
+    MlemModel,
+)
+from mlem.runtime.server import Server
+from mlem.ui import EMOJI_BUILD, EMOJI_OK, echo, set_offset
+
+from ..docker.base import (
+    DockerEnv,
+    DockerImage,
+    DockerImageBuilder,
+    DockerRegistry,
+    generate_docker_container_name,
+)
+from .context import K8sYamlBuildArgs, K8sYamlGenerator
+
+POD_STATE_MAPPING = {
+    "Pending": DeployStatus.STARTING,
+    "Running": DeployStatus.RUNNING,
+    "Succeeded": DeployStatus.STOPPED,
+    "Failed": DeployStatus.CRASHED,
+    "Unknown": DeployStatus.UNKNOWN,
+}
+
+
+class K8sDeploymentState(DeployState):
+    type: ClassVar = "kubernetes"
+    image: Optional[DockerImage]
+    pod_name: Optional[str]
+    kube_config_file_path: Optional[str] = None
+
+    def get_client(self):
+        config.load_kube_config(
+            config_file=self.kube_config_file_path
+            or os.getenv("KUBECONFIG", default="~/.kube/config")
+        )
+
+
+class K8sDeployment(MlemDeployment):
+    type: ClassVar = "kubernetes"
+    server: Optional[Server] = None
+    state_type: ClassVar = K8sDeploymentState
+    image_name: Optional[str] = None
+
+
+class K8sEnv(MlemEnv[K8sDeployment]):
+    type: ClassVar = "kubernetes"
+    deploy_type: ClassVar = K8sDeployment
+    registry: Optional[DockerRegistry] = DockerRegistry()
+
+    def create_k8s_resources():
+        pass
+
+    def deploy(self, meta: K8sDeployment):
+        self.check_type(meta)
+        redeploy = False
+        with meta.lock_state():
+            state: K8sDeploymentState = meta.get_state()
+            if state.image is None or meta.model_changed():
+                from ..docker.helpers import build_model_image
+
+                image_name = (
```

Comment:
move this block to separate method?

Reply:
hmm, I am following how it's done for Docker, so kinda keeping it if it's okay? Since Docker also follows the same structure.

Reply:
better to follow sagemaker since it's newer. we should refactor others the same way

Reply:
done